pennylane_lightning_gpu/src/simulator/StateVectorCudaMPI.hpp

```diff
@@ -941,6 +955,364 @@ class StateVectorCudaMPI
         return expect_val;
     }
 
+    /**
+     * @brief Scatter a CSR (Compress Sparse Row) format matrix.
+     *
+     * @tparam index_type Integer type used as indices of the sparse matrix.
+     * @param matrix CSR (Compress Sparse Row) format matrix.
+     * @param root Root rank of the scatter operation.
+     */
+    template <class index_type>
+    auto scatterCSRMatrix(std::vector<CSRMatrix<Precision, index_type>> &matrix,
+                          size_t root) -> CSRMatrix<Precision, index_type> {
+        // Bcast num_rows and num_cols
+        size_t local_num_rows;
+        size_t num_col_blocks;
+        std::vector<size_t> nnzs;
+
+        local_num_rows = size_t{1} << this->getNumLocalQubits();
+        num_col_blocks = mpi_manager_.getSize();
+
+        if (mpi_manager_.getRank() == root) {
+            for (size_t j = 0; j < matrix.size(); j++) {
+                nnzs.push_back(matrix[j].values.size());
+            }
+        }
+
+        size_t local_nnz = mpi_manager_.scatter<size_t>(nnzs, 0)[0];
+
+        CSRMatrix<Precision, index_type> localCSRMatrix(local_num_rows,
+                                                        local_nnz);
+
+        if (mpi_manager_.getRank() == root) {
+            localCSRMatrix.values = matrix[0].values;
+            localCSRMatrix.csrOffsets = matrix[0].csrOffsets;
+            localCSRMatrix.columns = matrix[0].columns;
+        }
+
+        for (size_t k = 1; k < num_col_blocks; k++) {
+            size_t dest = k;
+            size_t source = 0;
+
+            if (mpi_manager_.getRank() == 0 && matrix[k].values.size()) {
+                mpi_manager_.Send<std::complex<Precision>>(matrix[k].values,
+                                                           dest);
+            } else if (mpi_manager_.getRank() == k && local_nnz) {
+                mpi_manager_.Recv<std::complex<Precision>>(
+                    localCSRMatrix.values, source);
+            }
+
+            if (mpi_manager_.getRank() == 0 && matrix[k].values.size()) {
+                mpi_manager_.Send<index_type>(matrix[k].csrOffsets, dest);
+            } else if (mpi_manager_.getRank() == k && local_nnz) {
+                mpi_manager_.Recv<index_type>(localCSRMatrix.csrOffsets,
+                                              source);
+            }
+
+            if (mpi_manager_.getRank() == 0 && matrix[k].values.size()) {
+                mpi_manager_.Send<index_type>(matrix[k].columns, dest);
+            } else if (mpi_manager_.getRank() == k && local_nnz) {
+                mpi_manager_.Recv<index_type>(localCSRMatrix.columns, source);
+            }
```

Comment:
Can you add a description of what is happening in here?
Also, given the large numbers of if-else, is there another way to write this?

Reply:
Could we do 3 async sends (ISend) in the same if block, and then synchronous receives in the else block?

Reply:
Done!